system/lib/standalone/paths.c

```diff
@@ -0,0 +1,233 @@
+#define _GNU_SOURCE
+#include "paths.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sysexits.h>
+
+/// A name and file descriptor pair.
+typedef struct preopen {
+  /// The path prefix associated with the file descriptor.
+  const char* prefix;
+
+  /// The file descriptor.
+  __wasi_fd_t fd;
+} preopen;
+
+/// A simple growable array of `preopen`.
+static preopen* preopens;
+static size_t num_preopens;
+
+/// Are the `prefix_len` bytes pointed to by `prefix` a prefix of `path`?
+static bool
+prefix_matches(const char* prefix, size_t prefix_len, const char* path) {
+  // Allow an empty string as a prefix of any relative path.
+  if (path[0] != '/' && prefix_len == 0)
+    return true;
```

Comment:
IIRC out coding style uses braces around the body of single line `if` blocks (unless this is copied code in which case we should mention why the style is different in a comment up top.

Reply:
Yeah, this is wasi-libc derived code. I changed the brace style to match Emscripten's, as they aren't consistent about it anyway. Does this work for you?